### ISSUE TITLE
What to do when `$NINJA` points to an invalid ninja executable?

### ISSUE THREAD

**Issue by dnicolodi** [mesonbuild/meson-python#316]
I think something is off with the code that verifies the ninja version in `get_requires_for_build_wheel()` and when the `NINJA` environment variable is set: if the variable points to a path that does not exist or to an executable that reports a too old version, we ask the `ninja` package to be installed. In `build_wheel()` the `NINJA` environment variable is updated only if not set, thus it still points to the invalid `ninja` and Meson will fail when meson-python executes `meson setup`.

I think that failing when the `NINJA` environment variable is set incorrectly is the right thing to do. However, we could fail a bit sooner, in particular without installing the `ninja` package that is never going to be used.

**Comment by FFY00** [comment]
Yeah, that makes sense.

cc @henryiii

**Comment by henryiii** [comment]
Agree, it was an oversight, we should not request `ninja` if `NINJA` is set.

**Comment by eli-schwartz** [comment]
> In `build_wheel()` the `NINJA` environment variable is updated only if not set, thus it still points to the invalid `ninja`

Hmm, but why does it need to be set at all? The code being used to set it is the same code used to detect if it exists, which is also modeled on the same code Meson uses.

Specifically:
- if $NINJA is set, it is always a no-op:
  - and valid, it is setdefault'ed to the path it already is
  - and invalid, it is not setdefault'ed to anything
- if it is unset, it gets set to the detected version on $PATH, if any

The code only takes effect if the variable is unset. But in that case what you really care about is that meson running in the isolated env can pick up the relevant meson from either the system PATH or the pip-installed build_requires. Which Meson handles without setting this variable.

**Comment by dnicolodi** [comment]
I think it was "defensive coding" against the ninja discovery algorithm used in meson-python and the one used in Meson diverged. I agree that it is not necessary.

**Comment by eli-schwartz** [comment]
Hmm, fair enough :) it should be harmless either way and unrelated to this problem.

Perhaps _env_ninja_command() should check at the very bottom, right before falling out of the function with `return None`, if `env_ninja` is actually something. If it's something, but the version check failed, then raise an error instead of returning None.
